rpcserver.go

```diff
@@ -1607,20 +1616,264 @@ func validatePayReqExpiry(payReq *zpay32.Invoice) error {
 	return nil
 }
 
-// For each payment we need to know the msat amount, the destination
-// public key, and the payment hash.
+// payment is used by SendPayment, SendPaymentSync, SendToRoute
+// and SendToRouteSync.
+// When it is used by SendPayment / SendPaymentSync we need to know the
+// msat amount, the destination public key, and the payment hash.
+// When it is used by SendToRoute / SendToRouteSync we need to know the
+// routes and payment hash.
 type payment struct {
 	msat      lnwire.MilliSatoshi
 	dest      *btcec.PublicKey
 	pHash     []byte
 	cltvDelta uint16
+	routes    []*routing.Route
+}
+
+// paymentStream enables different types of payment stream, such as:
+// 		lnrpc.Lightning_SendPaymentServer
+// 		lnrpc.Lightning_SendToRouteServer
+// to execute sendPayment.
+type paymentStream struct {
+	recv func() (*payment, error)
+	send func(*lnrpc.SendResponse) error
 }
 
 // SendPayment dispatches a bi-directional streaming RPC for sending payments
 // through the Lightning Network. A single RPC invocation creates a persistent
 // bi-directional stream allowing clients to rapidly send payments through the
 // Lightning Network with a single persistent connection.
-func (r *rpcServer) SendPayment(paymentStream lnrpc.Lightning_SendPaymentServer) error {
+func (r *rpcServer) SendPayment(stream lnrpc.Lightning_SendPaymentServer) error {
+	return r.sendPayment(&paymentStream{
```

Comment:
I've tried your suggestion but ran into one problem:
- cannot pass a common interface for `stream` to  `sendPayment` because `lnrpc.Lightning_SendPaymentServer` and `lnrpc.Lightning_SendToRouteServer` are also an interface.